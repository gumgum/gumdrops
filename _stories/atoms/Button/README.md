## Props

The following props may be passed to configure the button. All additional props are passed through to the `<button>` element:

| name      | type      | description                                                                     | default   |
| --------- | --------- | ------------------------------------------------------------------------------- | --------- |
| className | `String`  | Additional class to be added to the outermost element                           |           |
| isBlock   | `Boolean` | Indicate if this is a block element button                                      | `false`   |
| isGroup   | `Boolean` | Indicate if this part of a button group                                         | `false`   |
| size      | `String`  | Indicate the size of the button. One of `xs`, `sm` or `lg`                      |           |
| context   | `String`  | Indicate the context of the button. Example `default`, `warning`, `danger` etc. | `default` |
