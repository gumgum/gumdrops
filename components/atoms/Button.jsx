import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

const Button = ({
    context,
    type,
    size,
    isGroup,
    group,
    isBlock,
    onClick,
    className,
    style,
    children,
    ...otherProps
}) => {
    const baseClass = 'gds-button';

    const rootClass = cx(baseClass, className, {
        [`${baseClass}--${context}`]: context,
        [`${baseClass}--${size}`]: size,
        [`${baseClass}-group__button`]: isGroup || group,
        [`${baseClass}--block`]: isBlock
    });

    return (
        <button className={rootClass} type={type} style={style} onClick={onClick} {...otherProps}>
            {children}
        </button>
    );
};

Button.displayName = 'Button';

Button.defaultProps = {
    context: 'default',
    type: 'button',
    isGroup: false,
    group: false,
    isBlock: false,
    onClick: null,
    style: {}
};

Button.propTypes = {
    /** outline, primary, secondary, success, warning, info, danger */
    context: PropTypes.string,
    type: PropTypes.string,
    /** xs, sm, lg */
    size: PropTypes.oneOf(['xs', 'sm', 'lg']),
    isGroup: PropTypes.bool,
    group: PropTypes.bool,
    isBlock: PropTypes.bool,
    onClick: PropTypes.func,
    className: PropTypes.string,
    style: PropTypes.object,
    children: PropTypes.node
};

export default Button;
