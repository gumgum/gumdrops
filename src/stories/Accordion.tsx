import React from 'react';
import cx from 'classnames';

interface AccordionCommonProps {
    color?: 'dark' | 'white';
    size?: 'sm' | '';
}

type OmitColor<T> = Omit<T, 'color'>;

export interface AccordionItemProps
    extends AccordionCommonProps,
        OmitColor<React.LiHTMLAttributes<HTMLLIElement>> {
    isOpen?: boolean;
    isLocked?: boolean;
    label: React.ReactChild;
}

export interface AccordionProps
    extends AccordionCommonProps,
        OmitColor<React.HTMLAttributes<Element>> {
    initialAllOpen?: boolean;
}

export const Accordion: React.FC<AccordionProps> = ({
    children,
    color,
    size = '',
    className,
    initialAllOpen,
    ...otherProps
}) => {
    const rootClass = cx('gds-accordion', className, {
        [`gds-accordion--${color}`]: color
    });

    const newChildren = React.Children.map(children, child => {
        if (!React.isValidElement<AccordionItemProps>(child)) {
            return child;
        }

        return React.cloneElement<AccordionItemProps>(child, {
            color,
            size,
            isOpen: child.props.isOpen ? child.props.isOpen : initialAllOpen
        });
    });

    return (
        <div className={rootClass} {...otherProps}>
            <ul className="gds-accordion-list">{newChildren}</ul>
        </div>
    );
};
