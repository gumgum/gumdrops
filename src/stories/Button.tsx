import React from 'react';
import cx from 'classnames';
import { Colors, Sizes } from 'types';

export interface ButtonProps extends React.HTMLAttributes<HTMLButtonElement> {
    color?: Colors;
    type?: 'button' | 'submit';
    size?: Sizes;
    isGroup?: boolean;
    isBlock?: boolean;
    disabled?: boolean;
}

const baseClass = 'gds-button';

export const Button: React.FC<ButtonProps> = ({
    color = 'default',
    size = 'md',
    type = 'button',
    isGroup,
    isBlock,
    children,
    disabled,
    className,
    ...otherProps
}) => {
    const rootClass = cx(baseClass, className, {
        [`${baseClass}--${color}`]: color,
        [`${baseClass}--${size}`]: size,
        [`${baseClass}-group__button`]: isGroup,
        [`${baseClass}--block`]: isBlock,
        '-disabled': disabled,
        '-pointer-events--none': disabled
    });

    return (
        <button className={rootClass} type={type} disabled={disabled} {...otherProps}>
            {children}
        </button>
    );
};
