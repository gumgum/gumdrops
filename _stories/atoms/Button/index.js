import React from 'react';
import { text, select, boolean, object } from '@storybook/addon-knobs';
import { optionalSelect } from '../../../components/utils/optionalSelect';
import { action } from '@storybook/addon-actions';

import readme from './README.md';
import Button from '../../../components/atoms/Button';

const sizeOptions = {
    xs: 'xs',
    sm: 'sm',
    lg: 'lg',
    'No Value': ''
};

const contextOptions = {
    default: 'default',
    outline: 'outline',
    primary: 'primary',
    secondary: 'secondary',
    success: 'success',
    warning: 'warning',
    info: 'info',
    danger: 'danger'
};

const component = () => (
    <Button
        type={text('Type', 'button')}
        size={optionalSelect('Size', sizeOptions, '')}
        context={select('Context', contextOptions, 'default')}
        isGroup={boolean('isGroup', false)}
        isBlock={boolean('isBlock', false)}
        onClick={action('button_clicked')}
        className={text('ClassName', '')}
        style={object('Style', {})}>
        <i className="fa fa-check -m-r-2" />
        Great Success
    </Button>
);

export default [readme, component];
