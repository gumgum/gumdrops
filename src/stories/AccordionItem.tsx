import React, { MouseEvent, KeyboardEvent, useState } from 'react';
import cx from 'classnames';
import { AccordionItemProps } from './Accordion';
import { keyCodes } from 'types';

type ToggleEvent = MouseEvent<HTMLLIElement> | KeyboardEvent<HTMLLIElement>;

export const AccordionItem: React.FC<AccordionItemProps> = ({
    isOpen: initialIsOpen,
    isLocked,
    label,
    size,
    color,
    className,
    children
}) => {
    const [isOpen, setIsOpen] = useState(initialIsOpen);

    const toggleOpen = (event: ToggleEvent): void => {
        event.stopPropagation(); // don't want nested accordions to propagate events
        event.preventDefault(); // prevents page scroll from space key
        const { type } = event;
        const { key } = event as KeyboardEvent;

        if (isLocked) return;
        const allowToggle =
            (type === 'keypress' && [keyCodes.space, keyCodes.enter].includes(key as keyCodes)) ||
            type === 'click';

        if (!allowToggle) return;

        setIsOpen(isOpen => !isOpen);
    };

    const rootClass = cx('gds-accordion__item', className, {
        [`gds-accordion__item--${color}`]: color,
        'gds-accordion__item--active': isOpen
    });

    const titleClass = cx('gds-accordion__item-title', {
        [`gds-accordion__item-title--${size}`]: size
    });

    const iconClass = cx('gds-accordion__item-icon', {
        [`gds-accordion__item-icon--${size}`]: size
    });

    const childClass = cx('gds-accordion__child-items', {
        [`gds-accordion__child-items--${size}`]: size
    });

    const newChildren = React.Children.map(children, child => {
        if (!React.isValidElement<AccordionItemProps>(child)) {
            return child;
        }

        return React.cloneElement<AccordionItemProps>(child, {
            color,
            size
        });
    });

    const regionId = `AccordionItem-region-${label}`;
    const labelId = `AccordionItem-label-${label}`;

    return (
        <li
            aria-pressed={isOpen}
            aria-controls={regionId}
            aria-expanded={isOpen}
            className={rootClass}
            onClick={toggleOpen}
            onKeyPress={toggleOpen}
            role="button"
            tabIndex={isLocked ? -1 : 0}>
            <h4 id={labelId} className={titleClass}>
                {label}
            </h4>
            {!isLocked && <span className={iconClass} />}
            <ul
                aria-labelledby={labelId}
                aria-hidden={!isOpen}
                className={childClass}
                id={regionId}
                role="region">
                {newChildren}
            </ul>
        </li>
    );
};
