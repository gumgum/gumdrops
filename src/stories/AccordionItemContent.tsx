import React from 'react';
import cx from 'classnames';
import { AccordionItemProps } from './Accordion';

export const AccordionItemContent: React.FC<Omit<AccordionItemProps, 'label'>> = ({
    className,
    color,
    size,
    children,
    ...otherProps
}) => {
    const stopClick: React.MouseEventHandler = e => {
        e.stopPropagation();
    };

    const rootClass = cx('gds-accordion__child-item', className, {
        [`gds-accordion__child-item--${color}`]: color
    });

    const titleClass = cx('gds-accordion__child-item-title', {
        [`gds-accordion__child-item-title--${size}`]: size
    });

    return (
        <li className={rootClass} onClick={stopClick} {...otherProps}>
            <h4 className={titleClass}>{children}</h4>
        </li>
    );
};
