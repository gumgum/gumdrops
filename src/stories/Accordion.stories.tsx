import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0';

import { Accordion, AccordionProps } from './Accordion';
import { AccordionItem } from './AccordionItem';
import { AccordionItemContent } from './AccordionItemContent';

export default {
    title: 'Accordion',
    component: Accordion
} as Meta;

const Template: Story<AccordionProps> = args => (
    <Accordion {...args}>
        <AccordionItem label="I'm locked and open" isLocked isOpen>
            <AccordionItemContent>Locked Content</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 1">
            <AccordionItemContent>Content 1</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 2">
            <AccordionItemContent>Content 2</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 3">
            <AccordionItemContent>Content 3</AccordionItemContent>
            <Accordion>
                <AccordionItem label="Nested Item">
                    <AccordionItemContent>Nested Content</AccordionItemContent>
                </AccordionItem>
            </Accordion>
        </AccordionItem>
    </Accordion>
);

export const Primary = Template.bind({});
Primary.args = {};
