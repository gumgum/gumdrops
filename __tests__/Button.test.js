import React from 'react';
import { render, screen } from '@testing-library/react';
import user from '@testing-library/user-event';
import { Button } from 'stories/Button';

describe('Button', () => {
    it('calls the given callback on click', () => {
        const mockOnClick = jest.fn();
        const { container } = render(<Button onClick={mockOnClick}>Test</Button>);
        user.click(screen.getByRole('button', { name: 'Test' }));
        expect(mockOnClick).toHaveBeenCalledTimes(1);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('clicking a disabled button does nothing', () => {
        const mockOnClick = jest.fn();
        render(
            <Button onClick={mockOnClick} disabled>
                Test
            </Button>
        );
        user.click(screen.getByRole('button', { name: 'Test' }));
        expect(mockOnClick).not.toHaveBeenCalled();
    });
});
