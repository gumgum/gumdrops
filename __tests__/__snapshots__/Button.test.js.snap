// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button calls the given callback on click 1`] = `
<button
  class="gds-button gds-button--default gds-button--md"
  type="button"
>
  Test
</button>
`;
